mp/src/game/client/momentum/ui/HUD/hud_speedometer_label.cpp

```diff
@@ -0,0 +1,329 @@
+{
+    m_flCurrentValue = 0.0f;
+    m_flPastValue = 0.0f;
+    m_flDiff = 0.0f;
+    m_flNextUpdateCheck = 0.0f;
+
+    m_bCustomDiff = false;
+    m_bDrawComparison = true;
+    m_bDoneFading = true;
+
+    m_pComparisonLabel->SetText("");
+    BaseClass::SetText("");
+
+    SetAlpha(HasFadeOutAnimation() ? 0 : 255);
+}
+
+void SpeedometerLabel::SetText(int value)
+{
+    char szValue[BUFSIZELOCL];
+    Q_snprintf(szValue, sizeof(szValue), "%i", value);
+    BaseClass::SetText(szValue);
+}
+
+void SpeedometerLabel::Update(float value)
+{
+    m_flCurrentValue = value;
+
+    ConvertUnits();
+    SetText(m_flCurrentValue);
+    Colorize();
+    m_bDoneFading = HasFadeOutAnimation() ? !StartFadeout() : false;
+}
+
+void SpeedometerLabel::SetUnitType(int type) 
+{ 
+    if (type >= SPEEDOMETER_UNITS_FIRST && type <= SPEEDOMETER_UNITS_LAST)
+        SetUnitType(SpeedometerUnits_t(type));
+}
+
+void SpeedometerLabel::ConvertUnits()
+{
+    if (m_eUnitType == SPEEDOMETER_UNITS_UPS)
+        return;
+
+    switch (m_eUnitType)
+    {
+    case SPEEDOMETER_UNITS_KMH:
+        m_flCurrentValue *= UPS_TO_KMH_FACTOR;
+        m_flDiff *= UPS_TO_KMH_FACTOR;
+        break;
+    case SPEEDOMETER_UNITS_MPH:
+        m_flCurrentValue *= UPS_TO_MPH_FACTOR;
+        m_flDiff *= UPS_TO_MPH_FACTOR;
+        break;
+    case SPEEDOMETER_UNITS_ENERGY:
+        // outlier as it requires more than just 1 velocity
+        if (m_bSupportsEnergyUnits)
+        {
+            const auto gravity = sv_gravity.GetFloat();
+            const auto pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+            Vector velocity = pPlayer->GetAbsVelocity();
+            const auto pEntData = pPlayer->GetCurrentUIEntData();
+            if (!pEntData)
+                break;
+
+            m_flCurrentValue =
+                (velocity.LengthSqr() / 2.0f + gravity * (pPlayer->GetLocalOrigin().z - pEntData->m_flLastJumpZPos)) /
+                gravity;
+        }
+        break;
+    default:
+        break;
+    }
+}
+
+void SpeedometerLabel::SetColorizeType(int type) 
+{
+    if (type >= SPEEDOMETER_COLORIZE_FIRST && type <= SPEEDOMETER_COLORIZE_LAST)
+        SetColorizeType(SpeedometerColorize_t(type));
+}
+
+void SpeedometerLabel::Colorize()
+{
+    if (m_flNextUpdateCheck > gpGlobals->curtime)
+        return;
+
+    switch (m_eColorizeType)
+    {
+    case SPEEDOMETER_COLORIZE_RANGE:
+        ColorizeRange();
+        break;
+    case SPEEDOMETER_COLORIZE_COMPARISON:
+        ColorizeComparison();
+        break;
+    case SPEEDOMETER_COLORIZE_COMPARISON_SEPARATE:
+        ColorizeComparisonSeparate();
+        break;
+    default:
+        SetFgColor(m_NormalColor);
+        break;
+    }
+
+    m_flPastValue = m_flCurrentValue;
+    m_flNextUpdateCheck = gpGlobals->curtime + MOM_COLORIZATION_CHECK_FREQUENCY;
+}
+
+void SpeedometerLabel::SetFadeOutAnimation(char *animationName, float *animationAlpha)
+{
+    m_pszAnimationName = animationName;
```

Comment:
You should `Q_strncpy` into a char array instead of just storing a pointer, and the check to see if it's valid is just checking if the first char is `\0`.